includes/class-wc-admin-reports-segmenting.php

```diff
@@ -0,0 +1,509 @@
+<?php
+/**
+ * Class for adding segmenting support without cluttering the data stores.
+ *
+ * @package  WooCommerce Admin/Classes
+ */
+
+defined( 'ABSPATH' ) || exit;
+
+/**
+ * Date & time interval and numeric range handling class for Reporting API.
+ */
+class WC_Admin_Reports_Segmenting {
+
+	/**
+	 * Array of all segment ids.
+	 *
+	 * @var array|bool
+	 */
+	protected $all_segment_ids = false;
+
+	/**
+	 * Query arguments supplied by the user for data store.
+	 *
+	 * @var array
+	 */
+	protected $query_args = '';
+
+	/**
+	 * SQL definition for each column.
+	 *
+	 * @var array
+	 */
+	protected $report_columns = array();
+
+	/**
+	 * WC_Admin_Reports_Segmenting constructor.
+	 *
+	 * @param array $query_args Query arguments supplied by the user for data store.
+	 * @param array $report_columns Report columns lookup from data store.
+	 */
+	public function __construct( $query_args, $report_columns ) {
+		$this->query_args     = $query_args;
+		$this->report_columns = $report_columns;
+	}
+
+	/**
+	 * Filters definitions for SELECT clauses based on query_args and joins them into one string usable in SELECT clause.
+	 *
+	 * @param array $columns_mapping Column name -> SQL statememt mapping.
+	 *
+	 * @return string to be used in SELECT clause statements.
+	 */
+	protected function prepare_selections( $columns_mapping ) {
+		if ( isset( $this->query_args['fields'] ) && is_array( $this->query_args['fields'] ) ) {
+			$keep = array();
+			foreach ( $this->query_args['fields'] as $field ) {
+				if ( isset( $columns_mapping[ $field ] ) ) {
+					$keep[ $field ] = $columns_mapping[ $field ];
+				}
+			}
+			$selections = implode( ', ', $keep );
+		} else {
+			$selections = implode( ', ', $columns_mapping );
+		}
+
+		if ( $selections ) {
+			$selections = ',' . $selections;
+		}
+
+		return $selections;
+	}
+
+	/**
+	 * Update row-level db result for segments in 'totals' section to the format used for output.
+	 *
+	 * @param array  $segments_db_result Results from the SQL db query for segmenting.
+	 * @param string $segment_dimension Name of column used for grouping the result.
+	 *
+	 * @return array Reformatted array.
+	 */
+	protected function reformat_totals_segments( $segments_db_result, $segment_dimension ) {
+		$segment_result = array();
+
+		if ( strpos( $segment_dimension, '.' ) ) {
+			$segment_dimension = substr( strstr( $segment_dimension, '.' ), 1 );
+		}
+
+		foreach ( $segments_db_result as $segment_data ) {
+			$segment_id = $segment_data[ $segment_dimension ];
+			unset( $segment_data[ $segment_dimension ] );
+			$segment_datum                 = array(
+				'segment_id' => $segment_id,
+				'subtotals'  => $segment_data,
+			);
+			$segment_result[ $segment_id ] = $segment_datum;
+		}
+
+		return $segment_result;
+	}
+
+	/**
```

Comment:
Thanks for these comment blocks, it really helped me grok what was going on here.